### PR TITLE
perf: optimize row merging

### PR DESCRIPTION
**Pull request by igorbernstein2**
This PR rewrites the row merging logic to be more correct and improve performance:
- extract row merging logic into its own class to simplify complexity of ReadRows handling
- Use OrderedDict instead of dict() for `{family: { qualifier: [] }}` data, this should maintain serverside ordering (family in creation order and qualifier in lexiographical). 
- define an explicit state machine with states implemented as methods
- add various optimizations like:
  - __slots__ on hot objects to avoid dict lookups
  - avoiding dict lookups for contiguous family and qualifier keys
 
Overall this improves performance by 20% and in my opinion is a lot more readable

### CODE REVIEW

File: google/cloud/bigtable/row_merger.py [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+        if self.row.cell.qualifier is None:
+            raise InvalidChunk("missing family for a new cell")
+
+        self.row.cell.timestamp = chunk.timestamp_micros
+        self.row.cell.labels = chunk.labels
+
+        if chunk.value_size > 0:
+            # explicitly avoid pre-allocation as it seems that bytearray
+            # concatenation performs better than slice copies.
+            self.row.cell.value = bytearray()
```
Comment (mutianf) [inline google/cloud/bigtable/row_merger.py:410]: how will we add the current chunk to the cell value? should we add self.row.cell.value += chunk.value to line 406?

File: google/cloud/bigtable/row_merger.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+    def _handle_cell_start(self, chunk):
+        # Ensure that all chunks after the first one either are missing a row
+        # key or the row is the same
+        if self.row.cells and chunk.row_key and chunk.row_key != self.row.row_key:
+            raise InvalidChunk("row key changed mid row")
+
+        if not self.row.cell:
+            self.row.cell = _PartialCell()
+
+        if chunk.HasField("family_name"):
```
Comment (mutianf) [inline google/cloud/bigtable/row_merger.py:393]: Can we simplify this to:

if !chunk.HasField("family_name") or !chunk.HasField("qualifier"):
  raise InvalidChunk("missing family_name or qualifier for a new cell")

?

File: google/cloud/bigtable/row_merger.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+            self.row.cell.family = chunk.family_name.value
+            self.row.cell.qualifier = None
+        if not self.row.cell.family:
+            raise InvalidChunk("missing family for a new cell")
+
+        if chunk.HasField("qualifier"):
+            self.row.cell.qualifier = chunk.qualifier.value
+        if self.row.cell.qualifier is None:
+            raise InvalidChunk("missing family for a new cell")
```
Comment (mutianf) [inline google/cloud/bigtable/row_merger.py:402]:
```suggestion
            raise InvalidChunk("missing qualifier for a new cell")
```

File: google/cloud/bigtable/row_merger.py (hunk 2 above)
Comment (mutianf) [reply google/cloud/bigtable/row_merger.py:393]: And should we move this check before we create a new PartialCell?

File: google/cloud/bigtable/row_merger.py [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+                raise InvalidChunk("In progress cell had a qualifier")
+            if chunk.timestamp_micros:
+                raise InvalidChunk("In progress cell had a timestamp")
+            if chunk.labels:
+                raise InvalidChunk("In progress cell had labels")
+
+        self.row.cell.value += chunk.value
+        self.row.cell.value_index += len(chunk.value)
+
+        if chunk.value_size > 0:
```
Comment (mutianf) [inline google/cloud/bigtable/row_merger.py:434]: what's chunk.value_size? Is it the remaining size of this cell?

File: google/cloud/bigtable/row_data.py [hunk 5]
```diff
@@ -513,125 +249,25 @@ def __iter__(self):
...
-                        cell.qualifier = previous.qualifier
+                self.last_scanned_row_key = self._row_merger.last_seen_row_key
+                self._counter += 1
+
+                yield row
+
+                if self._cancelled:
+                    break
+            # The last response might not have generated any rows, but it
+            # could've updated last_scanned_row_key
```
Comment (mutianf) [inline google/cloud/bigtable/row_data.py:269]: why would it be updated? Isn't last_seen_row_key only updated in handleRowComplete?

**Review by mutianf — CHANGES_REQUESTED** [review]

File: google/cloud/bigtable/row_merger.py (hunk 2 above)
Comment (mutianf) [reply google/cloud/bigtable/row_merger.py:393]: Add a comment to explain the family and qualifier buffer

**Review by mutianf — APPROVED** [review]

**Comment by Mariatta** [comment]
Can you add the documentation of row_merger to the Sphinx documentation?
First create `docs/row-merger.rst` (see the example content from `row-data.rst`)

Then add `row-merger` to the `usage.rst` line 18.

We should also update the usages of row_data.Cell / row_data.PartialRowData / row_data.InvalidChunk in the documentation to use the row_merger instead.

File: google/cloud/bigtable/row_merger.py [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+            self.row.cell.value = bytearray()
+            self.state = _State.CELL_IN_PROGRESS
+        else:
+            self.row.cell.value = chunk.value
+            self.state = _State.CELL_COMPLETE
+
+    def _handle_cell_in_progress(self, chunk):
+        # if this isn't the first cell chunk, make sure that everything except
+        # the value stayed constant.
+        if self.row.cell.value_index > 0:
```
Comment (dbolduc) [inline google/cloud/bigtable/row_merger.py:418]: can we drop `value_index` from `PartialCell` and just check for:
```py
if self.row.cell.value:
```

**Review by dbolduc — APPROVED** [review]
The row merging logic LGTM. I left some nits that you can fix or not.

File: google/cloud/bigtable/row_merger.py [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+                raise InvalidChunk("In progress cell had a qualifier")
+            if chunk.timestamp_micros:
+                raise InvalidChunk("In progress cell had a timestamp")
+            if chunk.labels:
+                raise InvalidChunk("In progress cell had labels")
+
+        self.row.cell.value += chunk.value
+        self.row.cell.value_index += len(chunk.value)
+
+        if chunk.value_size > 0:
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google/cloud/bigtable/row_merger.py:195]: yea, I wanted to keep explicit

File: google/cloud/bigtable/row_merger.py [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+    "in the column family {}. There are {} such cells."
+)
+
+
+class _State(Enum):
+    ROW_START = "ROW_START"
+    CELL_START = "CELL_START"
+    CELL_IN_PROGRESS = "CELL_IN_PROGRESS"
+    CELL_COMPLETE = "CELL_COMPLETE"
+    ROW_COMPLETE = "COMPLETE_ROW"
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google/cloud/bigtable/row_merger.py]: yep, thanks!

File: google/cloud/bigtable/row_merger.py [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+
+    def __init__(self, last_seen_row=b""):
+        self.last_seen_row_key = last_seen_row
+        self.state = _State.ROW_START
+        self.row = None
+
+    def process_chunks(self, response):
+        """
+        Process the chunks in the given response and yield logical rows.
+        The internal state of this class will maintain state across multiple
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google/cloud/bigtable/row_merger.py]: done

File: google/cloud/bigtable/row_merger.py [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+class _RowMerger(object):
+    """
+    State machine to merge chunks from a response stream into logical rows.
+
+    The implementation is a fairly linear state machine that is implemented as
+    a method for every state in the _State enum. In general the states flow
+    from top to bottom with some repetition. Each state handler will do some
+    sanity checks, update in progress data and set the next state.
+
+    There can be multiple state transitions for each chunk, ie. a single chunk
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google/cloud/bigtable/row_merger.py]: done

**Review by Mariatta — APPROVED** [review]

**Pull request merged**